File: solutions/go/transpose/1/transpose.go
package transpose

func computeTransposedLength(input []string) (int,[]int) {
    lengths := []int{}

    max := -1
    for i := len(input) - 1; i >= 0; i-- {
        if len(input[i]) > max {
            max = len(input[i])
        }
        lengths = append(lengths, max)
    }
    for i,j := 0, len(lengths) - 1; i < j; i,j = i + 1, j - 1 {
        lengths[i], lengths[j] = lengths[j], lengths[i]
    }
    return max, lengths
}

func Transpose(input []string) []string {
    max, lengths := computeTransposedLength(input)
    if max <= 0 {
        return []string{}
    }
    result := make([]string, max)
	for i, s := range input {
        for j, ch := range s {
            result[j] = result[j] + string(ch)
        }
        for j := len(s); j < lengths[i]; j++ {
			result[j] = result[j] + " "            
        }
    }
    return result
}

File: solutions/clojure/transpose/1/src/transpose.clj
(ns transpose
  (:require [clojure.string :as s]))

(defn max-len [s]
  (->> s
       (map count)
       (apply max)))

(defn pad [len char s]
  (->>  (concat s (repeat char))
        (take len)
        s/join))

(defn trim [{:keys [max result]} col]
  (let [trimmed (s/trimr col)
        size (count trimmed)]
    (if
        (> size max) {:max size :result (conj result trimmed)}
        {:max max :result (conj result (s/join (take max col)))})))

(defn transpose
  "Given a string, it returns the transposed version"
  [s]
  (let [s (s/split-lines s)
        len (max-len s)]
    (->> s
         (map (partial pad len \space))
         (apply mapv str)
         reverse
         (reduce trim {:max 0 :result '()})
         :result
         (s/join "\n"))))
